### PR TITLE
Enhancement/428/jinja2 template support

### PR DESCRIPTION
**Pull request by rexemin**
##### SUMMARY
Enhances the collection to add support for Jinja2 templates in `zos_copy` and `zos_job_submit`.

Addresses issues:
#428 
#487 
#488 
#698 
#699 
#710

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
zos_copy, zos_job_submit, and a new module util named template.

### CODE REVIEW

File: plugins/modules/zos_copy.py
Comment (richp405) [inline plugins/modules/zos_copy.py:2679]: Shouldn't this have a 'required' entry in it?

Comment (rexemin) [reply plugins/modules/zos_copy.py:2679]: Are you asking because of the possibility that a user may set options in `template_parameters` while ignoring `use_template`? It may need a `required` entry, I need to verify that

File: plugins/action/zos_copy.py
Comment (fernandofloresg) [inline plugins/action/zos_copy.py:165]: I think try block should only be needed when using template not in variable assignment. I suggest to use the block from line 172 to line 179.

File: plugins/action/zos_copy.py
Comment (fernandofloresg) [inline plugins/action/zos_copy.py:196]: same comment as above.

File: plugins/action/zos_job_submit.py
Comment (fernandofloresg) [inline plugins/action/zos_job_submit.py:41]: If result is a dict consider using .update() function when updating multiple lines. Debate me on this, is just a comment about visuals. You can ignore this

File: plugins/action/zos_job_submit.py
Comment (fernandofloresg) [inline plugins/action/zos_job_submit.py:113]: Same comment as in zos_copy action plugin.

File: plugins/doc_fragments/template.py
Comment (fernandofloresg) [inline plugins/doc_fragments/template.py:27]: Do you know what is the order of precedence when getting the variables from different source? Might be worth it adding into the docs.

File: plugins/doc_fragments/template.py
Comment (fernandofloresg) [inline plugins/doc_fragments/template.py:1]: Great that you found out how to import docs into other modules. Super useful!

File: plugins/module_utils/encode.py
Comment (fernandofloresg) [inline plugins/module_utils/encode.py:1]: Beware of this addition, I've seen this function being detected as missing of new addition in git prs. Just check this was not duplicated after merge.

File: plugins/doc_fragments/template.py
Comment (fernandofloresg) [inline plugins/doc_fragments/template.py:24]: Do you know which version of Jinja2 is supported? do we select it or are dependent on python version?

File: tests/functional/modules/test_zos_copy_func.py
Comment (fernandofloresg) [inline tests/functional/modules/test_zos_copy_func.py:1]: Do we need a test for copying with different encodings? I mean, does it even make sense, the controller most probably will be a unix/linux system with utf-8 or similar, and will update the template visually. But maybe if the file came from a different z/OS system with some strange codepage could make sense.

**Review by fernandofloresg — CHANGES_REQUESTED** [review]

File: plugins/doc_fragments/template.py
Comment (rexemin) [reply plugins/doc_fragments/template.py:27]: The[ Ansible documentation has the complete hierarchy](https://docs.ansible.com/ansible/latest/playbook_guide/playbooks_variables.html#variable-precedence-where-should-i-put-a-variable) the engine uses. I'll add a note about it and link it in the doc fragment.

File: plugins/doc_fragments/template.py
Comment (rexemin) [reply plugins/doc_fragments/template.py:24]: I don't know for sure which version of Jinja is installed by each version of Ansible, but since it's a dependency for the engine, we don't really control it.

File: plugins/modules/zos_copy.py
Comment (rexemin) [reply plugins/modules/zos_copy.py:2679]: I think I misunderstood your comment yesterday. If I remember correctly, arguments need to have either a `required` entry or a `default` one. By using `default` the engine infers `required=False`.

Comment (richp405) [reply plugins/modules/zos_copy.py:2679]: That's what I was asking about!  Okay, this looks okay to me.

**Review by richp405 — APPROVED** [review]

File: plugins/action/zos_job_submit.py
Comment (rexemin) [reply plugins/action/zos_job_submit.py:41]: I don't think I need to debate you, I browsed other modules and all of them use `update` when updating multiple items in a dictionary. Thanks for bringing it up!

File: tests/functional/modules/test_zos_copy_func.py
Comment (rexemin) [reply tests/functional/modules/test_zos_copy_func.py:1]: I could add a test for templates encoded in ISO8859-1, which is the other most common encoding.

Comment (fernandofloresg) [reply tests/functional/modules/test_zos_copy_func.py:1]: What do you think about this? @ketankelkar @ddimatos

**Review by AndreMarcel99 — COMMENTED** [review]
I really like the code is too clean and also I understand well how it works.

File: plugins/action/zos_copy.py
Comment (rexemin) [reply plugins/action/zos_copy.py]: I moved down the try block. It still includes the creation and assigment of the `TemplateRenderer` object because its `__init__` method can throw two exceptions.

Comment (fernandofloresg) [reply plugins/action/zos_copy.py]: Got it!

File: plugins/action/zos_copy.py
Comment (fernandofloresg) [reply plugins/action/zos_copy.py]: Not needed as "I moved down the try block. It still includes the creation and assigment of the TemplateRenderer object because its __init__ method can throw two exceptions."

**Review by fernandofloresg — APPROVED** [review]
Looks ok to me. Would have preferred to split it into multiple PRs, each for a module for easier review.

**Comment by rexemin** [comment]
Screenshot 2023-06-26 at 15 42 38
Screenshot 2023-06-26 at 15 42 56

File: plugins/module_utils/template.py
Comment (ddimatos) [inline plugins/module_utils/template.py:236]: Very small, is it `temp` , `tmp` just play it safe and say temporary. Python does have precedence rules for how it finds a temporary directory to write into so its not always /tmp because i have had to instruct users on how to override this so that our modules an write into a bigger directory, basically saying maybe play it safe and if it makes sense, just say 'temporary'

File: plugins/module_utils/template.py
Comment (ddimatos) [inline plugins/module_utils/template.py:293]: Same comment here regarding `temp`

Comment (ddimatos) [reply plugins/module_utils/template.py:293]: And just below this.

**Review by ddimatos — CHANGES_REQUESTED** [review]
@rexemin overall this really looks great, its quite clean the test cases were creative, this as new territory for us and you did a nice job. I did leave a few comments, most of them are simply doc and ok if want to pass on them but did have a couple of points /questions in the code you should have a look at and let me know if its something worth changing.

**Comment by ddimatos** [comment]
One additional comment, I don't think i saw any new return values coming back to the module, is there a need for any, such as `use_template` that might help us debug issues in the future? Maybe we don't need it.

File: plugins/action/zos_copy.py
Comment (rexemin) [reply plugins/action/zos_copy.py:213]: I'll add a cleanup just in case. The method can throw a variety of exceptions, but there are two that could appear while writing a new file and leave incomplete files.

File: plugins/action/zos_job_submit.py
Comment (rexemin) [reply plugins/action/zos_job_submit.py:132]: Yes, thanks for catching this.

**Comment by rexemin** [comment]
I don't think we need a new return value, the task will have the value for 'use_template' whenever the user adds `-vvv` or `-vvvv`

File: plugins/action/zos_copy.py
Comment (ddimatos) [reply plugins/action/zos_copy.py:215]: Thank you, always great to see your willingness to work with me.

File: plugins/action/zos_copy.py
Comment (ddimatos) [reply plugins/action/zos_copy.py]: Thank you

**Comment by ddimatos** [comment]
This is pretty much ready for merge (I resolved the ones I saw updated) but wondering on one unresolved comment from @fernandofloresg and then the other ones from myself if they are going to be passed on or corrected. @rexemin

**Review by ddimatos — APPROVED** [review]
Looks great, so excited to see this and its well done, thank you so much @rexemin

**Pull request merged**
